src/Mod/Part/App/PropertyTopoShapeList.cpp

```diff
@@ -0,0 +1,193 @@
+    for (unsigned int i = 0; i < lValue.size(); i++)
+        _lValueList[i] = lValue[i];
+    hasSetValue();
+}
+
+PyObject *PropertyTopoShapeList::getPyObject()
+{
+    PyObject* list = PyList_New(getSize());
+    for (int i = 0; i < getSize(); i++)
+        PyList_SetItem( list, i, _lValueList[i]->getPyObject());
+    return list;
+}
+
+
+void PropertyTopoShapeList::setPyObject(PyObject *value)
+{
+    // check container of this property to notify about changes
+    Part2DObject* part2d = dynamic_cast<Part2DObject*>(this->getContainer());
+
+    if (PySequence_Check(value)) {
+        Py_ssize_t nSize = PySequence_Size(value);
+        std::vector<TopoShape*> values;
+        values.resize(nSize);
+
+        for (Py_ssize_t i=0; i < nSize; ++i) {
+            PyObject* item = PySequence_GetItem(value, i);
+            if (!PyObject_TypeCheck(item, &(TopoShapePy::Type))) {
+                std::string error = std::string("types in list must be 'Shape', not ");
+                error += item->ob_type->tp_name;
+                throw Base::TypeError(error);
+            }
+
+            values[i] = static_cast<TopoShapePy*>(item)->getTopoShapePtr();
+        }
+
+        setValues(values);
+        if (part2d)
+            part2d->acceptGeometry();
+    }
+    else if (PyObject_TypeCheck(value, &(TopoShapePy::Type))) {
+        TopoShapePy  *pcObject = static_cast<TopoShapePy*>(value);
+        setValue(pcObject->getTopoShapePtr());
+        if (part2d)
+            part2d->acceptGeometry();
+    }
+    else {
+        std::string error = std::string("type must be 'Shape' or list of 'Shape', not ");
+        error += value->ob_type->tp_name;
+        throw Base::TypeError(error);
+    }
+}
+
+void PropertyTopoShapeList::Save(Writer &writer) const
+{
+    writer.Stream() << writer.ind() << "<ShapeList count=\"" << getSize() <<"\">" << endl;
+    writer.incInd();
+    for (int i = 0; i < getSize(); i++) {
+        _lValueList[i]->Save(writer);
+    }
+    writer.decInd();
+    writer.Stream() << writer.ind() << "</ShapeList>" << endl ;
+}
+
+void PropertyTopoShapeList::Restore(Base::XMLReader &reader)
+{
+    reader.readElement("ShapeList");
+    int count = reader.getAttributeAsInteger("count");
+    std::vector<TopoShape*> values;
+    values.reserve(count);
+    for (int i = 0; i < count; i++) {
+        reader.readElement("TopoShape");
+        std::string file (reader.getAttribute("file") );
+        if (!file.empty()) {
+            // add file to reader's list of files to be loaded by each newShape's RestoreDocFile()
+            TopoShape* newShape = new TopoShape();
+            values.push_back(newShape);
+            reader.addFile(file.c_str(), newShape);
+        }
+    }
+    reader.readEndElement("ShapeList");
+
+    // set the list to the restored values
+    setValues(std::move(values));
+}
+
+
+App::Property *PropertyTopoShapeList::Copy() const
+{
+    PropertyTopoShapeList *p = new PropertyTopoShapeList();
```

Comment:
Here you should consider creating deep copies of the shapes with BRepBuilderAPI_Copy. This is also done in PropertyPartShape::Copy()

Reply:
Done.